sklearn/metrics/pairwise.py

```diff
@@ -585,6 +592,119 @@ def cosine_distances(X, Y=None):
+    According the Gower formula, w is an attribute weight.
+
+    categorical_features: array-like, shape (n_features)
+    Indicates with True/False wheter a column is a categorical attribute.
+    This is useful when categorical atributes are represented as integer
+    values.
+
+    Returns
+    -------
+    similarities : ndarray, shape (n_samples, )
+
+    Notes
+    ------
+    Gower is a similarity measure for categorical, boolean and numerical mixed
+    data.
+
+    """
+
+    X, Y = check_pairwise_arrays(X, Y, dtype=(np.object, None)[issparse(X) or
+                                                               issparse(Y)])
+
+    rows, cols = X.shape
+
+    if categorical_features is None:
+        categorical_features = []
+        for col in range(cols):
+            if np.issubdtype(type(X[0, col]), np.number):
+                categorical_features.append(False)
+            else:
+                categorical_features.append(True)
+    # Calculates the normalized ranges and max values of numeric values
+    ranges_of_numeric = [0.0] * cols
+    max_of_numeric = [0.0] * cols
+    for col in range(cols):
+        if not categorical_features[col]:
+            max = None
+            min = None
+            if issparse(X):
+                col_array = X.getcol(col)
+                max = col_array.max() + 0.0
+                min = col_array.min() + 0.0
+            else:
+                col_array = X[:, col].astype(np.double)
+                max = np.nanmax(col_array)
+                min = np.nanmin(col_array)
+
+            if np.isnan(max):
+                max = 0.0
+            if np.isnan(min):
+                min = 0.0
+            max_of_numeric[col] = max
+            ranges_of_numeric[col] = (1 - min / max) if (max != 0) else 0.0
+
+    if w is None:
+        w = [1] * cols
+
+    yrows, ycols = Y.shape
+
+    dm = np.zeros((rows, yrows), dtype=np.double)
+
+    for i in range(0, rows):
+        j_start = i
+
+        # for non square results
+        if rows != yrows:
+            j_start = 0
+
+        for j in range(j_start, yrows):
+            sum_sij = 0.0
+            sum_wij = 0.0
+            for col in range(cols):
+                value_xi = X[i, col]
+                value_xj = Y[j, col]
+
+                if not categorical_features[col]:
+                    if (max_of_numeric[col] != 0):
+                        value_xi = value_xi / max_of_numeric[col]
+                        value_xj = value_xj / max_of_numeric[col]
+                    else:
+                        value_xi = 0
+                        value_xj = 0
+
+                    if ranges_of_numeric[col] != 0:
+                        sij = abs(value_xi - value_xj) / ranges_of_numeric[col]
+                    else:
+                        sij = 0
+                    wij = (w[col], 0)[np.isnan(value_xi) or np.isnan(value_xj)]
+                else:
+                    sij = (1.0, 0.0)[value_xi == value_xj]
```

Comment:
This indexing hack is unreadable. Please avoid it.

Reply:
removed